src/encode_append.rs

```diff
@@ -59,68 +93,70 @@ fn extract_length_data(data: &[u8], input_len: usize) -> Result<(u32, usize, usi
 	Ok((new_len, encoded_len, encoded_new_len))
 }
 
-impl<T: Encode + 'static> EncodeAppend for Vec<T> {
-	type Item = T;
-
-	fn append<'a, I: IntoIterator<Item=&'a Self::Item>>(
-		mut self_encoded: Vec<u8>,
-		iter: I,
-	) -> Result<Vec<u8>, Error> where Self::Item: 'a, I::IntoIter: ExactSizeIterator {
-		let iter = iter.into_iter();
-		let input_len = iter.len();
-
-		// No data present, just encode the given input data.
-		if self_encoded.is_empty() {
-			crate::codec::compact_encode_len_to(&mut self_encoded, iter.len())?;
-			iter.for_each(|e| e.encode_to(&mut self_encoded));
-			return Ok(self_encoded);
-		}
+// Item must have same encoding as encoded value in the encoded vec.
+fn append_or_new_vec_with_any_item<'a, Item, I>(
+	mut self_encoded: Vec<u8>,
+	iter: I,
+) -> Result<Vec<u8>, Error>
+where
+	Item: 'a + Encode,
+	I: IntoIterator<Item = &'a Item>,
+	I::IntoIter: ExactSizeIterator,
+{
+	let iter = iter.into_iter();
+	let input_len = iter.len();
+
+	// No data present, just encode the given input data.
+	if self_encoded.is_empty() {
+		crate::codec::compact_encode_len_to(&mut self_encoded, iter.len())?;
+		iter.for_each(|e| e.encode_to(&mut self_encoded));
+		return Ok(self_encoded);
+	}
 
-		let (new_len, encoded_len, encoded_new_len) = extract_length_data(&self_encoded, input_len)?;
+	let (new_len, encoded_len, encoded_new_len) = extract_length_data(&self_encoded, input_len)?;
 
-		let replace_len = |dest: &mut Vec<u8>| {
-			Compact(new_len).using_encoded(|e| {
-				dest[..encoded_new_len].copy_from_slice(e);
-			})
-		};
+	let replace_len = |dest: &mut Vec<u8>| {
+		Compact(new_len).using_encoded(|e| {
+			dest[..encoded_new_len].copy_from_slice(e);
+		})
+	};
 
-		let append_new_elems = |dest: &mut Vec<u8>| iter.for_each(|a| a.encode_to(dest));
+	let append_new_elems = |dest: &mut Vec<u8>| iter.for_each(|a| a.encode_to(dest));
 
-		// If old and new encoded len is equal, we don't need to copy the
-		// already encoded data.
-		if encoded_len == encoded_new_len {
-			replace_len(&mut self_encoded);
-			append_new_elems(&mut self_encoded);
+	// If old and new encoded len is equal, we don't need to copy the
+	// already encoded data.
+	if encoded_len == encoded_new_len {
+		replace_len(&mut self_encoded);
+		append_new_elems(&mut self_encoded);
 
-			Ok(self_encoded)
-		} else {
-			let size = encoded_new_len + self_encoded.len() - encoded_len;
+		Ok(self_encoded)
+	} else {
+		let size = encoded_new_len + self_encoded.len() - encoded_len;
 
-			let mut res = Vec::with_capacity(size + input_len * mem::size_of::<T>());
-			unsafe { res.set_len(size); }
+		let mut res = Vec::with_capacity(size + input_len * mem::size_of::<Item>());
```

Comment:
here is the only change of current behavior the hint size can be different but this is fine.